### DISCUSSION TITLE
Can virtua cause memory leak on frequent scroll

### DISCUSSION THREAD

**Discussion by morajabi** [inokawa/virtua#258] [Q&A] [answered]
In our chat app we're hunting for a memory leak. I noticed this memory uptick on frequently scrolling (around 100 MB) which doesn't go down. I'm still not 100% sure this is virtua, but wanted to open this discussion to understand better if it can be indeed caused by virtua. 

https://github.com/inokawa/virtua/assets/12202757/d7f5b92f-1980-4398-8a0e-8a2746922996

Higher quality:
[CleanShot 2023-11-26 at 18.22.11.mp4.zip](https://github.com/inokawa/virtua/files/13467176/CleanShot.2023-11-26.at.18.22.11.mp4.zip)

**Comment by inokawa** [comment, marked as answer]
I hope it is not caused by virtua. In theory it will not cause memory leak but there are several things that might help:

- If props.children is ReactNode, it is flattened to array and keeped with useMemo.
- Items' sizes and offsets are cached in number array. The length will match total length of items.
- Item's actual DOM is keeped in WeakMap to associate DOM and item's index until unmount.
- Items are observed with ResizeObserver on mount and unobserved on unmount.
